lib/OnyxCache.js

```diff
@@ -0,0 +1,183 @@
+     * @param {function(): Promise<string[]>} fallback - a callback to retrieve keys from storage
+     * @returns {Promise<string[]>}
+     */
+    getAllKeys(fallback) {
+        // When we have storage keys in cache resolve right away
+        if (this.storageKeys.size > 0) {
+            return Promise.resolve([...this.storageKeys]);
+        }
+
+        // Read keys using the fallback. This task optimizes concurrent calls see `resolveTask` for details
+        const startTask = () => fallback().then((keys) => {
+            this.storageKeys = new Set(keys);
+            return keys;
+        });
+
+        return this.resolveTask('getAllKeys', startTask);
+    }
+
+    /**
+     * Get a cached value from storage
+     * Provide a promise returning fallback that will be executed when
+     * the value is not available in cache
+     * @template T
+     * @param {string} key
+     * @param {function(key: string): Promise<T>} fallback - a fallback to retrieve the item from storage
+     * @returns {Promise<T>}
+     */
+    getValue(key, fallback) {
+        // When we have cache resolve right away
+        if (this.hasCacheForKey(key)) {
+            return Promise.resolve(this.storageMap[key]);
+        }
+
+        // Get the value using the fallback. This task optimizes concurrent calls see `resolveTask` for details
+        const startTask = () => fallback(key).then(value => this.update(key, value));
+        return this.resolveTask(`getValue:${key}`, startTask);
+    }
+
+    /**
+     * Check whether cache has data for the given key
+     * @param {string} key
+     * @returns {boolean}
+     */
+    hasCacheForKey(key) {
+        return isDefined(this.storageMap[key]);
+    }
+
+    /**
+     * Update or add data to cache
+     * Providing value is optional. Skipping value would just capture the key so that
+     * it's available in `getAllKeys`
+     * @param {string} key
+     * @param {*} [value]
+     * @returns {*} value - returns the cache value
+     */
+    update(key, value) {
+        // Keep storage keys up to date so that `getAllKeys` works correctly
+        this.storageKeys.add(key);
+
+        // When a 2nd parameter (value) is provided - store it under the key
+        if (isDefined(value)) {
+            this.storageMap[key] = value;
+        }
+
+        return value;
+    }
+
+    /**
+     * Remove data from cache
+     * @param {string} key
+     */
+    remove(key) {
+        this.storageKeys.delete(key);
+        delete this.storageMap[key];
+    }
+
+    /**
+     * Merge data to cache, any non existing keys will be crated
+     * Array items are concatenated
+     * Object items are merged
+     * If the existing value in cache and the new value are not the same type
+     * the existing value is replaced by the new value
+     * In case the merge value for a key is undefined - the key is removed from cache
+     * @param {Array<[string, *]>} pairs - array of key value pairs
+     */
+    merge(pairs) {
```

Comment:
I wonder if passing `pairs` is the way to go here. Seems like it could be cleaner to just work with one `key/value` at a time and would also allow us to return a single value that we can call `keyChanged()` with after.

Reply:
Merge works with an object now